### PR TITLE
[resource host+process] Allow host name and process owner attributes to be excluded

### PR DESCRIPTION
**Pull request by flaviocdc**
## Changes
Process owner and Host name may contain private data, so there has to be a way to not include them as part of the telemetry record.

In order to achieve this, I'm adding an extra parameter to `AddHostDetector` and `AddProcessDetector` that allows consumers on the detectors to control whether or not this data will be added to the `Resource`.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry.Resources.Host/HostResourceBuilderExtensions.cs [hunk 1]
```diff
@@ -17,8 +17,19 @@ public static class HostResourceBuilderExtensions
     /// <param name="builder">The <see cref="ResourceBuilder"/> being configured.</param>
     /// <returns>The instance of <see cref="ResourceBuilder"/> being configured.</returns>
     public static ResourceBuilder AddHostDetector(this ResourceBuilder builder)
+    {
+        return AddHostDetector(builder, true);
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Resources.Host/HostResourceBuilderExtensions.cs:21]: nit: Slightly better readability IMO:

```suggestion
        return AddHostDetector(builder, includeHostName: true);
```

File: src/OpenTelemetry.Resources.Process/ProcessResourceBuilderExtensions.cs [hunk 2]
```diff
@@ -17,8 +17,19 @@ public static class ProcessResourceBuilderExtensions
     /// <param name="builder">The <see cref="ResourceBuilder"/> being configured.</param>
     /// <returns>The instance of <see cref="ResourceBuilder"/> being configured.</returns>
     public static ResourceBuilder AddProcessDetector(this ResourceBuilder builder)
+    {
+        return AddProcessDetector(builder, true);
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Resources.Process/ProcessResourceBuilderExtensions.cs:21]: nit: Slightly better readability IMO:

```suggestion
        return AddProcessDetector(builder, includeProcessOwner: true);
```

**Review by CodeBlanch — APPROVED** [review]
LGTM

**Comment by CodeBlanch** [comment]
Related: https://github.com/open-telemetry/opentelemetry-specification/issues/4218

**Comment by Kielek** [comment]
Note for myself from https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/#recommended:

> Instrumentations SHOULD add the attribute by default if it’s readily available and can be [efficiently populated](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/#performance-suggestions). Instrumentations MAY offer a configuration option to disable Recommended attributes.

> 
> Instrumentations that decide not to populate Recommended attributes due to [performance](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/#performance-suggestions), security, privacy, or other consideration by default, SHOULD allow for users to opt-in to emit them as defined for the Opt-In requirement level (if the attributes are logically applicable)

**Comment by Kielek** [comment]
There is couple types of the resources
 - Required - always should be available
 - Conditionally required - always should be available
 - Recommended - it MAY be possibility to opt-out
 - Opt-in - it may be possibility to opt-in
 
[Host resource detector sem conv](https://github.com/open-telemetry/semantic-conventions/blob/v1.27.0/docs/resource/host.md) contains 7 recommended and 2 Opt-in attributes. Not all of them are implemented here.

I think that we need better to expose this publicly (not on the merge level, it can be on adding method). Adding ~9 boolean attributes is not good idea IMO.

Maybe we should have
Somethin like:
`public static ResourceBuilder AddProcessDetector(this ResourceBuilder builder, IEnumerable<string> optOutAttributes, Enumerable<string> optInAttributes)`
or 
`public static ResourceBuilder AddProcessDetector(this ResourceBuilder builder, IEnumerable<HostOptOutAttribute> optOutAttributes, Enumerable<HostOptInAttribute> optInAttributes)`?

If operate on strings - what should happen when someone want to disable Required attribute or is passing attribute not supported by the given resource detectot?

It will be great to give it also possibility to configure it from env vars to make it AutoInstrumentation friendly.

**Comment by CodeBlanch** [comment]
@Kielek Personally I'm not a fan of APIs which rely on magic strings. Could we do something more strongly typed?

```csharp
public static class HostResourceBuilderExtensions
{
    public static ResourceBuilder AddHostDetector(this ResourceBuilder builder) {}
    public static ResourceBuilder AddHostDetector(this ResourceBuilder builder, Action<HostResourceOptions> configure) {}
}

public sealed class HostResourceOptions
{
    internal HostResourceOptions(IConfiguration configuration)
    {
        var config = configuration.GetValue("OTEL_DOTNET_HOST_RESOURCE_ATTRIBUTES", (string?)null);
        if (config != null)
        {
            // TODO: Parse config and apply onto bools (somehow)
        }
    }

    #region Recommended (on by default)
    public bool IncludeArchitecture { get; set; } = true;

    public bool IncludeId { get; set; } = true;

    public bool IncludeImageId { get; set; } = true;

    public bool IncludeImageName { get; set; } = true;

    public bool IncludeImageVersion { get; set; } = true;

    public bool IncludeName { get; set; } = true;

    public bool IncludeType { get; set; } = true;
    #endregion

    #region Opt-In (off by default)
    public bool IncludeIpAddress { get; set; }

    public bool IncludeMacAddress { get; set; }
    #endregion
}
```

**Comment by Kielek** [comment]
@flaviocdc, could you please follow @CodeBlanch description? At least for already implemented resources?
Please update also README.md with example how to use this option.

**Comment by ysolomchenko** [comment]
I agree with the idea of using configurations through options. I suggest doing the same for the remaining resources, and it's worth creating an issue for this

**Pull request closed without merging**